README.md
# edly
Electron diffraction simulator web dashboard.

![](static/images/edly.png)

## Installation
Run the `install.sh` file to complete installation.

## Miscellaneous
### datasets
Experimental dataset (raw frames) are located in the `database` folder. Locally available datasets can manually be put in this directory.

### Clear a session info
Automatically cleared every day but can also be cleared manually with :
```
rm -rf static/data/tmp/<session_id>
```


changelog.md
# Version log
## 0.0.8dev
### Thu  7 Dec 08:13:13 GMT 2023
- added `load_pets` in bloch.py
- initial addition of `rock_frames` to keep track of exp frames in rocking curve
- custom calculation of excitation error from `ED_importer`
- test_custom
- ED_version  : 3930006
### Wed 15 Nov 14:25:08 GMT 2023
- mathjax update formula
- issue#67 pixel range :
  - added frame_folder flag to pets/dials for nxy
- ED version:exp 7245da9
### Wed 18 Oct 16:07:18 BST 2023
- issue#64 : rocking curve
  - fix slight integration issue
  - save currently simulated rocking curve dialog prompt `dialog_save_rock`
  - `rock_name` click location changed
- tests :
  - dialog save rock new test
- remove some mathjax stuffs

### Tue 17 Oct 10:57:29 BST 2023
- initialization bug fixes : changes to `init_bloch_panel`
- tests : added *test_switch_structure*

## 0.0.7
### Fri 13 Oct 13:01:43 CEST 2023
- tests :
  - test *solve_rock.py*
  - fix pb `test_download` and `test_zenodo` when run with other tests
  - added possibility to check the style of an element
  - changed `elt.clear()` to execute_script as it would delete `input_rock_name`
  - all tests passing with `--sleep=1`
  - put short timers at some points but there are still failures when rocking on full speed mode
### Thu 12 Oct 18:30:11 CEST 2023
- tests:
  - capturing browser console to the report folder *console.log*
### Thu 12 Oct 13:11:11 BST 2023
- issue#62 : delete frames
  - simple new feature

### Wed 11 Oct 19:06:36 BST 2023
- issue#63 : cancel download
  - added ids to relevant html DOM
  - link is updated only when clicking back on the set_link button
  - added `download.downloading` in *mainCtrl.js*
  - download jobs have an ID
- test :
  - added download/cancel download and manipulate zenodo entries
- update README

### Mon  9 Oct 20:45:54 BST 2023
- added test option to run only specific files as `run_tests.sh -m 'frames'`
- login and close are run no matter what
- browser creation is a global fixture defined in *conftest.py*
- removed *test_base.py*

### Thu  5 Oct 17:26:24 BST2023
- reports of tests sent by email

### Wed 27 Sep 13:31:28 BST 2023
- added test_frames to automatically test frame import and navigation

### Mon 25 Sep 17:07:56 BST 2023
- issue#61 :
  - fix empty database error at init
- ccp4ED version : `exp 7759eca`

### Tue 19 Sep 17:39:04 BST 2023
- issue#55 :
  - fixed create new project errors

### Tue 18 Jul 17:56:01 BST 2023
- rocking curve :
  - delete option
  - get info on load change
  - new_rock

### Tue 18 Jul 14:09:27 BST 2023
- issue#52 : fix pets import
- added rock_frames to old simus
- ED version : exp 7759eca

### Thu  6 Jul 09:08:32 BST 2023
- issue#51
  - enter simu number messes up the frame number
  - full rock reflection selection in Io vs Ic
  - `show_rock` display both exp and simu rocking curve
  - changed exp integration to only cover simulated frames
  - save `rock_frames` info into `Rock` object to retrieve at anytime
  - fixed automatic rock_name selection at startup
  - button sync frame and simu nb
  - green saved button
### Fri 30 Jun 17:50:41 BST 2023
- issue#49 : rock curves and IovsIc bugs
  - fix update rock curves bug when in IovsIc mode
  - properly update IovsIc curve when changing thickness
  - reflections index shown on IovsIc
  - clickable reflections on IovsIc graph
  - fix available  graphs bugs at refresh
- removed structure links in navigation
- tests :
  - added code coverage support

### Wed 28 Jun 16:34:17 BST 2023
- bug fixes :
  - remove try except `init_structure()`
  - removed trailing '/' in replace database_path
- tests:
  - run_test uses getopts with simplified flag options
  - added report,dev,lvl flags to run_tests
  - added level markers to tests

### Tue 27 Jun 21:11:03 BST 2023
- bug fixes :
  - check pets key before loading it
  - check `df_G` in dict for `beam_vs_thick`
  - update `session['crys']` when importing `cif_file`
  - fix popup not showing for some buttons
  - increased popup timer to allow smooth button selection
- tests :
  - added lots of ids
  - added selenium_utils files for function callbacks
  - new passing test :  `import dat`,
  - new passing test `frames_mode`(click_nav, key_nav, input_nav, br, heat)
  - new passing test `bloch_mode` (single,thick,thicks,rock_solve,save, rock_curves, integrate)
- note :
  - app does not display fine in offline mode
  - version : electron-diffraction commit : exp 347da7c

### Mon 26 Jun 20:15:58 BST 2023
- bug fixes :
  - do not display bloch or frames if not available
  - do not display import options if no structure is selected
  - update display if deleting structure was the active structure
  - full frames folder path removing database path
  - create database link and folder if not present
  - added submit functions to all dialog forms
  - fix ng-style bug on delete_dialog element
- tests :
  - added ids to lots of widgets
  - failing in headless mode
  - added `.send_keys()`/`.click()` for `focus`/`upload_file` interactions
  - new test :  `new_structure` passing
  - new test :  `import_cif`    passing
  - new test :  `import_frames` passing
  - new test :  `delete_struct` passing

### Mon 26 Jun 10:15:37 BST 2023
- enhancement:
   - added create and delete structure buttons
   - display stuff
   - full link for frames folder
   - confirmation delete dialog
   - ng-blur and ng-focus on search input boxes
### Mon 26 Jun 08:47:56 BST 2023
- bug fix :  
  - fixed folder name in `data_path`
- cosmetic :
  - `fetch` flag when updating zenodo so it only does it on cloud request
  - minor display stuff

### Wed 21 Jun 14:55:31 BST 2023
- issue#44 import info:
  - added update zenodo functionality although does not work in headless
  - added zenodo.py file
  - not fully tested
- enhancement :
  - added search frames in local database functionality
  - switch from local to zenodo database
  - click to edit frames link (default non editable)
- bug fixes :
  - fix dl_state stall and extracting msg

bw_app.py
# import importlib as imp
from subprocess import check_output,Popen,PIPE
import json,tifffile,os,sys,glob,time,datetime,crystals #,base64,hashlib
from flask import Flask,Blueprint,request,url_for,redirect,jsonify,session,render_template
from functools import wraps
import numpy as np,pandas as pd
from EDutils import utilities as ut             #;imp.reload(ut)
from EDutils import felix as fe                 #;imp.reload(fe)
from utils import displayStandards as dsp
from utils import glob_colors as colors
import plotly.express as px
import plotly.graph_objects as go
from blochwave import bloch
from blochwave import bloch_pp as bl            #;imp.reload(bl)
from in_out import*
bw_app = Blueprint('bw_app', __name__)

felix_data = {}


def login_required(f):
    @wraps(f)
    def wrap(*args, **kwargs):
        # print(session.get('logged_in'))
        if session.get('logged_in')==True:
            return f(*args, **kwargs)
        else:
            # flash("You need to login first")
            return redirect(url_for('login'))
    return wrap



######################################################
#### Structure realted
######################################################
@bw_app.route('/upload_cif', methods=['POST'])
def upload_cif():
    try:
        f = request.files['file_cif']
        cif_file=os.path.join(session['path'],f.filename)
        f.save(cif_file)
        return 'ok'
    except Exception as e:
        msg='could not import cif file:'
        print(colors.red+msg)
        print(colors.magenta,e,colors.black)
        return "%s : %s" %(msg, e.__str__())


@bw_app.route('/new_structure', methods=['POST'])
def new_structure():
    data = request.form
    mol = data['name']
    val = data['val']

    path=mol_path(mol)
    cif_file=os.path.join(path,data[val])
    if val=='cif':cif_file+='.cif'
    elif val=='pdb':cif_file+='.pdb'
    # elif val=='file':cif_file.replace('.cif','.struct.cif')

    msg = 'cif file issue'
    if not mol :#or mol=="new":
        msg='Choose a proper name for the structure'
    else:
        if not os.path.exists(path):
            try:
                check_output('mkdir %s' %path,shell=True)
                if   val=='cif' :
                    crys = crystals.Crystal.from_database(data['cif'])
                    ut.crys2felix(crys,opt='w',out=cif_file)
                    cif_file=data[val]
                elif val=='pdb' :
                    crys = crystals.Crystal.from_pdb(data['pdb'],download_dir='.')
                    check_output('mv pdb%s.ent %s' %(data['pdb'],cif_file),shell=True)
                elif val=='file':
                    cif_path=os.path.join(session['path'],data['file'])
                    if os.path.exists(cif_path):
                        check_output('cp %s %s' %(cif_path,cif_file),shell=True)

                #check sucessful import
                if cif_file:
                    session['mol']=mol
                    init_session()
                    msg='ok'
            except Exception as e:
                check_output('rm -rf %s' %path,shell=True)
                raise Exception(e)
                msg=e.__str__()
        else:
            msg='%s already exists' %mol
    print(msg)
    return msg

@bw_app.route('/set_structure', methods=['POST'])
def set_structure():
    data = request.form
    # data=json.loads(request.data.decode())
    # print(data)
    session['mol']=data['mol']
    init_mol()
    return session['mol']


################################################
#### frame
################################################
@bw_app.route('/get_frame', methods=['POST'])
def get_frame():
    data=json.loads(request.data.decode())
    zmax  = data['zmax']
    frame = int(data['frame'])
    exp_img,sim_img = ['static/images/dummy.png']*2
    if session['dat']['sim']:
        sim_img = get_img_frame(frame,zmax['sim'],'sim')
    if session['dat']['exp']:
        exp_img = get_img_frame(frame,zmax['exp'],'exp')
    session['frame']=frame  #;print(frame)
    return json.dumps({'exp':exp_img,'sim':sim_img})

@bw_app.route('/update_zmax', methods=['POST'])
def update_zmax():
    data=json.loads(request.data.decode())
    img=get_img_frame(session['frame'],data['zmax'],data['key'])
    session['zm_counter']+=1
    return img

def get_img_frame(frame,zmax,key):
    session['modes']['analysis']='frames'
    offset = ''
    if key=='sim':
        frame=min(max(1,frame-session['sim']['offset']),session['sim']['max_frame'])
        offset='(%d frames offset )' %session['sim']['offset']
    frame_str=str(frame).zfill(session[key]['pad'])
    png_file=png_path(session['path'],'%s_%s' %(key,frame_str))
    if not os.path.exists(png_file) or not zmax==session[key]['z_max'][frame-1]:
        tiff_file=get_path(session['mol'],key,frame_str)
        im = tifffile.imread(tiff_file)
        dsp.stddisp(im=[im],caxis=[0,zmax],cmap='viridis',
            figsize=(10,)*2,
            pOpt="p", axPos=[0.05,0.05,0.9,0.9],
            name=png_file,title='frame %s %s' %(frame_str,offset), opt='sc')

        session[key]['zmax'] = zmax
        session[key]['z_max'][frame-1] = zmax

    ##### Dirty fix for automatic reload
    # The image is stored as `png_file`. It is copied and sent back
    # to angularJS as `new_tmp` to force reloading
    tmp=session[key]['tmp']
    old_tmp=png_path(session['path'],'tmp_%s_%d' %(key,tmp))
    new_tmp=png_path(session['path'],'tmp_%s_%d' %(key,tmp+1))
    p=Popen('rm %s;cp %s %s' %(old_tmp,png_file,new_tmp),shell=True,stdout=PIPE,stderr=PIPE)
    o,e=p.communicate()
    if e:print(e.decode())
    session[key]['tmp']+=1
    # print(session[key]['tmp'])
    return new_tmp



########################
#### structure related
########################
@bw_app.route('/set_mode', methods=['POST'])
def set_mode():
    data = json.loads(request.data.decode())
    # print(data)
    key  = data['key']
    session['modes'][key] = data['val']
    session['mol']  = session['mol']
    return json.dumps({key:session['modes'][key]})

@bw_app.route('/set_visible', methods=['POST'])
def set_visible():
    data=json.loads(request.data.decode())
    key = data['key']
    session['time']=time.time()
    session['vis'][key]=data['v']
    return json.dumps({key:session['vis'][key]})


############################################################################
#### Init
############################################################################
@login_required
@bw_app.route('/init', methods=['GET'])
def init():
    now = time.time()
    print('username : ' ,session['username'])
    if session.get('id') and os.path.exists(session.get('path')):
        init_mol()
        if (now-session['last_time'])>24*3600:
            print('warning:session created at %s has expired ' %session['last_time'])
            print(check_output('rm -rf %s' %session.get('path'),shell=True).decode())
            # init_args()
    else:
        id = '%s_%s' %(session['username'],create_id())
        session_path=os.path.join('static','data','tmp',id)
        print(check_output('mkdir -p %s' %session_path,shell=True).decode())
        print(colors.green+'creating new session %s' %id+colors.black)
        session['path'] = session_path
        session['id']   = id
        session['b0_path'] = get_pkl(session['id'])
        if not session.get('mol'):session['mol'] = 'glycine'
        init_session()

    # if session['dat']['felix'] and not session['mol'] in felix_data.keys():
    #     if os.path.exists(felix_pkl(session['mol'])):
    #         felix_data[session['mol']] = fe.load_felix(session['path'])
    #     else:
    #         felix_data[session['mol']] = fe.Felix(session['path'],session['mol'])



    # print('sending init info')
    info=['mol','dat','frame','crys','cif_file','mode']
    session_data = {k:session[k] for k in info}
    # frames : exp,sim
    session_data['zmax'] = {}
    session_data['max_frame'] = 0
    if session['sim']:
        session_data['max_frame']   = session['sim']['max_frame']
        session_data['zmax']['sim'] = session['sim']['zmax']
    if session['exp']:
        session_data['max_frame']   = max(session['exp']['max_frame'],session_data['max_frame'])
        session_data['zmax']['exp'] = session['exp']['zmax']
    # print(session_data['max_frame'])
    # session_data['structures'] = [s for s in structures if s!=session['mol']]
    # session_data['gifs'] = gifs
    return json.dumps(session_data)

def init_session():
    init_mol()
    # init_args()

def init_mol():
    mol = session['mol']
    exp = get_frames(mol,'exp')
    sim = get_frames(mol,'sim',d={'offset':10})
    dat = {
        'exp':type(exp)==dict,
        'sim':type(sim)==dict,
        'pets':os.path.exists(os.path.join(mol_path(mol),'pets')),
        'felix':os.path.exists(os.path.join(mol_path(mol),'reflprofiles_strong.dat')),
        }
    # if not os.path.exists(session['b0_path']) or new:
    struct_files = glob.glob(os.path.join(mol_path(mol),'*.cif'))
    if len(struct_files):
        struct_file=struct_files[0]
        base_file = os.path.basename(struct_file)[:-4]
        ##crystals library is rubbish as saved cif messes up the symmetry
        if base_file in crystals.Crystal.builtins:struct_file=base_file
    else:
        struct_file = glob.glob(os.path.join(mol_path(mol),'*.pdb'))[0]
    # elif os.path.basename(struct_file)[-3:]=='pdb':
    # print(struct_file)


    b0=bloch.Bloch(struct_file,path=session['path'],name='b',solve=False)
    # b0=ut.load_pkl(session['b0_path'])

    crys,cif_file=b0.crys,b0.cif_file
    crys_dat = {'file':os.path.basename(cif_file)}
    crys_dat.update({k:b_str(crys.__dict__[k],2) for k in ['a1', 'a2', 'a3']})
    crys_dat.update(dict(zip(['a','b','c','alpha','beta','gamma'],
        b_str(crys.lattice_parameters,2).split(',') )))

    if not session.get('modes'):session['modes']={'analysis':'bloch'}
    session['modes']['manual'] = not dat['pets']
    if not session.get('frame'):session['frame'] = 1

    now = time.time()
    session['cif_file'] = cif_file
    session['crys']     = crys_dat
    session['dat'] = dat
    session['sim'] = sim
    session['exp'] = exp
    session['zm_counter'] = 0 #dummy variable
    session['last_time']  = now
    session['time']  = now
    session['mode']  = 'bloch'


def get_frames(mol,dat,d={}):
    frames = glob.glob(get_path(mol,dat,'*'))
    max_frame=len(frames)
    frames_dict=None
    if max_frame>0:
        pad = len(os.path.basename(frames[0]).replace('.tiff',''))
        frames_dict = {
            'tmp':0,'zmax':50, 'z_max':[50]*max_frame,
            'max_frame':max_frame,'pad':pad
        }
        frames_dict.update(d)
    return frames_dict
